cmd/kubeadm/app/cmd/images.go

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"runtime"
+	"strings"
+
+	"github.com/spf13/cobra"
+	flag "github.com/spf13/pflag"
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	kubeadmapiext "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha1"
+	"k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	"k8s.io/kubernetes/cmd/kubeadm/app/features"
+	"k8s.io/kubernetes/cmd/kubeadm/app/images"
+	"k8s.io/kubernetes/cmd/kubeadm/app/phases/addons/dns"
+	kubeadmutil "k8s.io/kubernetes/cmd/kubeadm/app/util"
+	configutil "k8s.io/kubernetes/cmd/kubeadm/app/util/config"
+)
+
+// NewCmdImages returns the "kubeadm images" command
+func NewCmdImages(out io.Writer) *cobra.Command {
+	cfg := &kubeadmapiext.MasterConfiguration{}
+	kubeadmapiext.SetDefaults_MasterConfiguration(cfg)
+	var cfgPath, featureGatesString string
+	var err error
+
+	cmd := &cobra.Command{
+		Use:   "list-images",
+		Short: "Print a list of images kubeadm will use. The configuration file is used in case any images or image repositories are customized.",
+		Run: func(_ *cobra.Command, _ []string) {
+			if cfg.FeatureGates, err = features.NewFeatureGate(&features.InitFeatureGates, featureGatesString); err != nil {
+				kubeadmutil.CheckErr(err)
+			}
+			i, err := NewImages(cfgPath, cfg)
+			kubeadmutil.CheckErr(err)
+			kubeadmutil.CheckErr(i.Run(out))
+		},
+	}
+	AddImagesConfigFlag(cmd.PersistentFlags(), cfg, &featureGatesString)
+	AddImagesFlags(cmd.PersistentFlags(), &cfgPath)
+
+	return cmd
+}
+
+// NewImages returns a "kubeadm images" command
+func NewImages(cfgPath string, cfg *kubeadmapiext.MasterConfiguration) (*Images, error) {
+	internalcfg, err := configutil.ConfigFileAndDefaultsToInternalConfig(cfgPath, cfg)
+	if err != nil {
+		return nil, fmt.Errorf("could not convert cfg to an internal cfg: %v", err)
+	}
+
+	return &Images{
+		cfg: internalcfg,
+	}, nil
+}
+
+// Images defines the struct used for "kubeadm images"
+type Images struct {
+	cfg *kubeadmapi.MasterConfiguration
+}
+
+// Run runs the images command and writes the result to the io.Writer passed in
+func (i *Images) Run(out io.Writer) error {
+	imgs := []string{}
+	imgs = append(imgs, images.GetCoreImage(constants.KubeAPIServer, i.cfg.ImageRepository, i.cfg.KubernetesVersion, i.cfg.UnifiedControlPlaneImage))
```

Comment:
can't we wrap this into a utility f(n)?  images.GetAllImages()

Reply:
We certainly can. I usually leave that for when it's required, but I'm happy to refactor that if you think it's useful.